src/condor_utils/dc_coroutines.h

```diff
@@ -292,10 +261,204 @@ namespace dc {
+
+			std::suspend_always final_suspend() noexcept {
+				if( exception ) { std::rethrow_exception( exception ); }
+				return {};
+			}
+
+			void unhandled_exception() { exception = std::current_exception(); }
+
+			struct awaiter : std::suspend_always {
+				awaiter(promise_type * p) : promise(p) { }
+
+				R await_resume() {
+					return promise->response;
+				}
+
+				private:
+					promise_type * promise;
+			};
+
+			awaiter yield_value( const T & from ) {
+				value = from;
+				return {this};
+			}
+
+			void return_value( const T & from ) {
+				value = from;
+			}
+		};
+
+		handle_type handle;
+
+		Piperator() : handle(nullptr) {}
+		Piperator(handle_type h) : handle(h) {}
+
+		// Rule of 5, part I.
+		~Piperator() { if( handle ) { handle.destroy(); } }
+
+		// Rule of 5, part II.
+		Piperator( const Piperator & other ) = delete;
+
+		// Rule of 5, part III.
+		Piperator & operator = ( const Piperator & other ) = delete;
+
+		//
+		// Rule of 5, part IV.
+		//
+		// A std::coroutine_handle<> is trivially (shallow) copyable, but if
+		// we don't unset the other object's handle, it will destroy
+		// that handle when the other object goes out of scope.
+		//
+		// Since we're getting the handle's promise, we don't need to worry
+		// about copying the current yield or return value; the promise is
+		// storing that, and will be unchanged by these operations.
+		//
+		Piperator( Piperator && other ) {
+			this->full = other.full;
+
+			if( this->handle ) { handle.destroy(); }
+			this->handle = other.handle;
+			other.handle = nullptr;
+		}
+
+		// Rule of 5, part V.
+		Piperator & operator = ( Piperator && other ) {
+			this->full = other.full;
+
+			if( this->handle ) { handle.destroy(); }
+			this->handle = other.handle;
+			other.handle = nullptr;
+
+			return * this;
+		}
+
+
+		//
+		// Consider changing this so it's harder to accidentally resume
+		// the coroutine before setting the response.  Right now the
+		// required loop is:
+		//
+		//      do {
+		//          value = the_coroutine();
+		//          the_coroutine.handle.promise().response = next;
+		//      } while(! the_coroutine.handle.done());
+		//
+		// which could clearly be improved by an operator() that set
+		// the "response", and by a function on the_coroutine which
+		// means that we wouldn't have to reach inside it for this
+		// loop to work.  (Which may just mean skipping the fill() call.)
+		//
+		explicit operator bool() {
+			fill();
+			return (! handle.done());
+		}
+
+		T operator()() {
+			fill();
+			full = false;
+			return std::move(handle.promise().value);
+		}
+
+		void set_co_yield_value( const R &r ) {
+			handle.promise().response = r;
+		}
+
+		private:
+			bool full = false;
+
+			void fill() {
+				if(! full) {
+					handle();
+					if( handle.promise().exception ) {
+						std::rethrow_exception(handle.promise().exception);
+					}
+					full = true;
+				}
+			}
+	};
+
+} // end namespace cr
```

Comment:
@GregThain The new coroutine code.  Tested in `test_generator.cpp` because I wrote a simpler generator first.

Reply:
Resolving because of review comments elsewhere.